scapy/contrib/automotive/uds.py

```diff
@@ -1006,3 +1006,60 @@ def __init__(self, sock, pkt=UDS() / UDS_TP(), interval=2):
             interval: interval between two packets
         """
         PeriodicSenderThread.__init__(self, sock, pkt, interval)
+
+
+def UDS_SessionEnumerator(sock):
+    """ Enumerates session ID's 0x0 - 0xff
+        and returns list of UDS()/UDS_DSC() packets
+        with valid session types
+
+    Args:
+        sock: socket where packets are sent
+    """
+    pkts = [UDS()/UDS_DSC(diagnosticSessionType=x) for x in range(0xff)]
+    results = []
+    for p in pkts:
+        a = sock.sr1(p, timeout=0.5, verbose=False)
+        if a != None:
+            if a.service == 0x50 or \
+                    a.negativeResponseCode not in [0x10, 0x11, 0x12]:
+                results.append(p)
+        sock.send(UDS() / UDS_ER())
+        time.sleep(0.5)
+    return results
+
+
+def UDS_ServiceEnumerator(sock, session="DefaultSession"):
+    """ Enumerates every service ID
+        and returns list of tuples. Each tuple contains
+        the session and the respective positive response
+
+    Args:
+        sock: socket where packet is sent periodically
+        session: session in which the services are enumerated
+    """
+    pkts = [UDS(service=x) for x in set(x & ~0x40 for x in range(0x100))]
```

Comment:
Scapy supports packet generators. For instance:

```python 
a = IP(dst=["www.google.com", "www.google.fr"])
sr(a)
```
You can also use tuples with the format (start, end) to generate packets: `TCP(dport=(1,255))`.
Those packet generators can then be iterated through, which is done automatically when sending.

Please use this, rather than creating the packets yourself, then calling `sr1` plenty of times, each time creating tons of lists. Also applies above

Reply:
I changed it for the UDS_ServiceEnumerator function. 
In UDS_SessionEnumerator it's not possible because I need to send a UDS_ER() after 
every packet.